### PR TITLE
[external-assets] Normalize assets defs in InternalAssetGraph

### PR DESCRIPTION
**Pull request by smackesey**
## Summary & Motivation

Move the site of normalization of the assets definitions in a repo to the `InternalAssetGraph`. Normalization is: source asset conversion + dependency relative asset key resolution + generation of missing assets def for referenced assets.

A previous PR introduced conversion of source assets to external assets at the `RepositoryDataBuilder` level. However, we couldn't completely expunge `SourceAsset` here because repositories have a public API that returns `SourceAsset`-- so we needed to keep around the `SourceAssets` passed into the repo, while only passing their external asset representation into our internals.

This external asset representation of `SourceAsset` was merged into the `assets_defs_by_key` map for the repo. Though we have not heard any reports of problems from this, this was technically a breaking change, as `assets_defs_by_key` would now return not only the assets defs a user provided to a repo, but also the external asset representation of source assets.

This PR improves on this situation in several ways:

- It reverts the change to `RepositoryDefinition.assets_defs_by_key`. After this PR, this method will return only the `AssetsDefinition` passed in by the user. This is consistent with `source_assets_by_key`.
- It moves asset normalization logic  into `AssetGraph`. The `AssetGraph` cached on a `RepositoryDefinition` becomes the source of truth for us internally-- internal pathways operate on the normalized set of assets obtained from the `AssetGraph`. This creates a clean separation between our public exposure of the definitions our `RepositoryDefinition` takes in and their private internal representation.
- A bonus is that our tests can keep using `AssetGraph.from_assets` instead of going through a repository in order to get access to realistic behavior (as there is no normalization stage skipped when going through `from_assets`).

Most of the changes in the PR are just changing internal callsites of `RepositoryDefinition.assets_defs_by_key` to `RepositoryDefinition.asset_graph.assets_defs`.

## How I Tested These Changes

Existing test suite.

### CODE REVIEW

**Comment by smackesey** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/20223?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#20223** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/20223?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#20222** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/20222?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#20221** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/20221?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#20140** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/20140?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 1 other dependent PR ([#20114](https://github.com/dagster-io/dagster/pull/20114) <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/20114?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#20145** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/20145?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#20186** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/20186?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#20129** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/20129?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#20118** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/20118?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#20189** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/20189?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @smackesey and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: python_modules/dagster/dagster/_core/definitions/internal_asset_graph.py [hunk 1]
```diff
@@ -39,16 +41,64 @@ def __init__(
         }
 
     @staticmethod
-    def from_assets(
-        all_assets: Iterable[Union[AssetsDefinition, SourceAsset]],
-        asset_checks: Optional[Sequence[AssetChecksDefinition]] = None,
-    ) -> "InternalAssetGraph":
-        from dagster._core.definitions.external_asset import create_external_asset_from_source_asset
+    def normalize_assets(
```
Comment (smackesey) [inline python_modules/dagster/dagster/_core/definitions/internal_asset_graph.py:44]: This just encapsulates logic moved from `RepositoryDataBuilder`.

File: python_modules/dagster/dagster_tests/asset_defs_tests/test_assets.py [hunk 2]
```diff
@@ -1566,12 +1566,12 @@ def load_input(self, context):
 def test_context_assets_def():
     @asset
     def a(context):
-        assert context.assets_def == a
```
Comment (schrockn) [inline python_modules/dagster/dagster_tests/asset_defs_tests/test_assets.py:1569]: why this change?

File: python_modules/dagster/dagster_tests/core_tests/execution_tests/test_asset_backfill.py [hunk 3]
```diff
@@ -710,7 +710,7 @@ def external_asset_graph_from_assets_by_repo_name(
 
         external_asset_nodes = external_asset_nodes_from_defs(
             repo.get_all_jobs(),
-            repo.assets_defs_by_key,
+            {k: ad for ad in repo.asset_graph.assets_defs for k in ad.keys},
```
Comment (schrockn) [inline python_modules/dagster/dagster_tests/core_tests/execution_tests/test_asset_backfill.py:713]: why not leave `assets_defs_by_key` as a useful shortcut? This seems unnecessarily thrashy?

**Review by schrockn — CHANGES_REQUESTED** [review]
Pushing this into the AssetGraph makes sense but I don't think you should eliminate `assets_defs_by_key` from `RepositoryDefinition` as it is convenient and makes this diff unnecessarily thrashy.

E.g. going from

```python
   return AssetValueLoader(self.assets_defs_by_key, instance=instance)
```

to 

```python
        assets_defs_by_key = {k: ad for ad in self.asset_graph.assets_defs for k in ad.keys}
        return AssetValueLoader(assets_defs_by_key, instance=instance)
```

is a regression

File: python_modules/dagster/dagster_tests/asset_defs_tests/test_assets.py (hunk 2 above)
Comment (smackesey) [reply python_modules/dagster/dagster_tests/asset_defs_tests/test_assets.py:1569]: Normalization of an assets def changes the object identity due to the `with_attributes` call when resolving relative AssetKey deps, so checks against `AssetsDefinition` object identity need to change (not something we generally rely on, which is why it's only this test).

**Comment by smackesey** [comment]
>I don't think you should eliminate assets_defs_by_key from RepositoryDefinition as it is convenient and makes this diff unnecessarily thrashy.

`assets_defs_by_key` wasn't deleted (`external_assets_defs_by_key` was, but this was recently added and is called in only one place). From the summary:

>It reverts the change to RepositoryDefinition.assets_defs_by_key. After this PR, this method will return only the AssetsDefinition passed in by the user. This is consistent with source_assets_by_key.

So `assets_defs_by_key` now returns exactly what the user passed in. This was the behavior up until the PR a week or so ago that introduced source asset conversion in the `RepositoryDataBuilder`. When that PR landed, `assets_defs_by_key` started including the external assets from those converted source assets.

Since the change is reverted here, we can't use `assets_defs_by_key` internally because our internal callsites need the _normalized_ assets defs, which must be sourced from the asset graph (`assets_defs_by_key` does not have them because its contents are pre-normalization).

**Comment by schrockn** [comment]
> Since the change is reverted here, we can't use RepositoryDefinition.assets_defs_by_key internally because our internal callsites need the normalized assets defs, which must be sourced from the asset graph (assets_defs_by_key does not have them because its contents are intentionally pre-normalization).

I think this will cause a lot of confusion. How are our own engineers supposed to know to not use this property internally?

**Comment by smackesey** [comment]
>I think this will cause a lot of confusion. How are our own engineers supposed to know to not use this property internally?

The situation is a little unfortunate but I think unavoidable. I don't think we should be injecting auto-created definitions into the public `RepositoryDefinition.assets_defs_by_key`-- it will potentially confuse users who use this method together with `source_assets_by_key`.

I can add some comments explaining the situation, but if we want access to provide an internal API for a normalized set of assets (which I think we absolutely do, otherwise you have to perform ad hoc normalization on site, like resolving relative references inside `build_assets_job`), then IMO it needs to live elsewhere than `RepositoryDefinition.assets_defs_by_key`. Also after this PR + rest of stack the codebase is going to have `AssetGraph` everywhere so `AssetGraph.assets_defs` is what people will mostly see internally.

**Comment by schrockn** [comment]
Based on zoom conversation I support this. We need `assets_defs_by_keys` on `RepositoryDefinition` to directly return what was passed in, rather than canonicalized source assets.

Please add a helper property to to `AssetGraph` of `assets_defs_by_keys` to that we can just directly use that when needed, because some of the thrash here is a usability regression.

**Comment by schrockn** [comment]
Cool makes sense to me

**Comment by smackesey** [comment]
>Cool makes sense to me

This PR should be good to go then

**Review by schrockn — APPROVED** [review]
